### PR TITLE
Hollow Knight:  fix split cloak for reverse fill

### PR DESCRIPTION
**Pull request by Berserker66**
## What is this fixing or adding?
I have no idea

## How was this tested?
running generate once

## Actual

I found this bit of suspicious code, that I don't really know what it does. But whatever it does do, would need replication in reverse in remove to work correctly. So whatever it is doing it is doing wrong.

### CODE REVIEW

**Comment by Berserker66** [comment]
@dewiniaid any comments as you wrote it?

**Comment by Alchav** [comment]
Did you run generate with split mothwing cloak on

**Comment by Berserker66** [comment]
no

**Comment by Alchav** [comment]
When you get two Left_Mothwing_Cloaks or two Right_Mothwing_Cloaks, you get the Shadewing Cloak upgrade for that direction and, if you got one of the other directional Mothwing Cloaks, then you also get Shadewing Cloak for that direction as well. So if you have one of one direction and two of the other, it must collect two LEFTDASH and two RIGHTDASH effects

**Comment by Berserker66** [comment]
pushed what I believe to be an implementation of your description. Also tested genering with split cloak and it generated.

**Comment by Alchav** [comment]
This replaces 9 lines of code with 26. If it works and this is how you want it to work then I don't care either way, but I don't see why bother

**Comment by Berserker66** [comment]
because in reverse fill, the original solution breaks apart and assumes the upgraded dash is available, when it is in fact not.

**Comment by ThePhar** [comment]
image

Bruh.

File: worlds/hk/__init__.py [hunk 1]
```diff
@@ -514,23 +514,40 @@ def create_vanilla_location(self, location: str, item: Item):
...
-        if item.name in {"Left_Mothwing_Cloak", "Right_Mothwing_Cloak"}:
-            if state.prog_items.get(('RIGHTDASH', item.player), 0) and \
-                    state.prog_items.get(('LEFTDASH', item.player), 0):
-                (state.prog_items["RIGHTDASH", item.player], state.prog_items["LEFTDASH", item.player]) = \
-                    ([max(state.prog_items["RIGHTDASH", item.player], state.prog_items["LEFTDASH", item.player])] * 2)
+                prog_items[effect_name, self.player] += effect_value
+
+            # a directional overflow in dash grants an upgrade for the other side
+            if item.name == "Right_Mothwing_Cloak" and \
+                    prog_items["RIGHTDASH", self.player] > 2 > prog_items["LEFTDASH", self.player]:
```
Comment (Alchav) [inline worlds/hk/__init__.py:523]: I don't get why it's checking for greater than 2?

Comment (Berserker66) [reply worlds/hk/__init__.py:523]: Because, the way I understood it, is that when that direction can't be upgraded anymore it upgrades the other instead. And there's 2 upgrades per direction.

Comment (BadMagic100) [reply worlds/hk/__init__.py:523]: Yeah, I don't think this implementation is quite correct

So, deep diving into ItemChanger we can find this implementation for split cloak: https://github.com/homothetyhk/HollowKnight.ItemChanger/blob/master/ItemChanger/Resources/items.json#L5582
https://github.com/homothetyhk/HollowKnight.ItemChanger/blob/master/ItemChanger/Tags/ItemTreeTag.cs

This means that if left cloak is unobtained, you get one left dash. If left cloak has already been obtained, it will be shade cloak (either left or full depending if you have right dash or not). After that, the item tree is exhausted and you would actually get nothing. The same is true for right dash. Since AP makes 2 left and 2 right, I believe it should be the case that precisely the 4th dash you obtain would just do nothing.

Tldr I think instead of overflowing you'd just no-op in such cases. But I'd need to verify the behavior is as I described above in an actual AP game

Comment (Exempt-Medic) [reply worlds/hk/__init__.py:523]: I know that if you make an HK AP game, you end up with 3 Mothwing Cloaks; 2 in one direction, 1 in the other. I'm not sure why/how it makes 3 instead of 4, but there is [code to convert the Shade Cloak randomly into a Left or Right Mothwing Cloak](https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/hk/__init__.py#L294-L298) based on your settings

Comment (BadMagic100) [reply worlds/hk/__init__.py:523]: Oh, I actually read that code and i completely missed the if/randint (so in a normal game, without usage of getitem, there should only be 3)

So actually I think this overflowing logic just doesn't need to exist at all

Comment (BadMagic100) [reply worlds/hk/__init__.py:523]: Ok, I finally remembered to go double check on this. here is what I found

* 3 dashes get placed in current AP. This is all well and good. One left, one right, one which is either left or right at random
* The correct handling should be something like the following:
  ```python
  has_left_dash = prog_items["LEFTDASH", self.player]
  has_right_dash = prog_items["RIGHTDASH", self.player]
  has_any_shade_cloak = prog_items["LEFTDASH", self.player] > 1 or prog_items["RIGHTDASH", self.player] > 1
  if item.name == "Left_Mothwing_Cloak":
    # we always get 1 leftdash - then we'll go see if we need any special handling to upgrade
    # to shade cloak
    prog_items["LEFTDASH", self.player] += 1
    if has_left_dash and has_right_dash:
      # we already had both directional dashes, so we want right shade cloak too
      prog_items["RIGHTDASH", self.player] += 1  
    elif has_any_shade_cloak:
      # we already had right shade cloak so make this our left shade cloak too
      prog_items["LEFTDASH", self.player] += 1
  ```

In the code above, I've not included any handling for dupes because they don't exist. Mirroring this logic for right dash and inverting the logic for reverse fill (which I understand is the goal of this PR) is left as an exercise for the reader

**Review by BadMagic100 — CHANGES_REQUESTED** [review]

**Comment by BadMagic100** [comment]
I would say it's accurate to say (in its current iteration at least) that this PR is aiming to fix split cloak for reverse fill; I'd recommend renaming the PR accordingly and updating the description

**Comment by BadMagic100** [comment]
Sometimes between when I originally remarked on the titling and when the title got updated, I discovered this method is not actually called in fill_restrictive, so that would put it back to the drawing board what it is actually trying to do - it's very clear that remove should mirror add, but not why I should care

**Comment by Alchav** [comment]
> Sometimes between when I originally remarked on the titling and when the title got updated, I discovered this method is not actually called in fill_restrictive, so that would put it back to the drawing board what it is actually trying to do - it's very clear that remove should mirror add, but not why I should care

Yeah pretty sure it is not usually called during fill. There is a state.remove or something to uncollect, but I don't think it's used much, and it isn't going to uncollect anything that was swept up using the item you've removed.

**Comment by BadMagic100** [comment]
I took the time last night to verify that World.remove is only called via CollectionState.remove, which is in turn only called while calculating playthrough. So this is probably needed for accuracy of spoiler log.

**Comment by BadMagic100** [comment]
Closing this as the test it was thought to be blocking was merged (see referencing PRs above). There is certainly a bug but it's of minimal consequence (may cause incorrect playthrough, doesn't affect generation). Additionally, the approach here to fix it is incorrect and there has not had forward progress for >2 years. This issue will be addressed in the HK rewrite
